arch/x86/core/intel64/locore.S

```diff
@@ -0,0 +1,586 @@
+1:
+#endif
+
+	/* FIXME: ...disabling the PICs is common to IA32 and Intel64... */
+
+#ifdef CONFIG_PIC_DISABLE
+	movb $0xff, %al
+	outb %al, $0x21
+	outb %al, $0xA1
+#endif
+
+	/* next, switch to our own GDT/IDT and stack. */
+
+	lgdt gdt48
+	lidt idt48
+	jmpl $X86_KERNEL_CS_32, $1f
+1:	movw $X86_KERNEL_DS_32, %ax
+	movw %ax, %ds
+	movw %ax, %ss
+	movw %ax, %es
+	movw %ax, %fs
+	movw %ax, %gs
+
+	movl $(exception_stack + CONFIG_EXCEPTION_STACK_SIZE), %esp
+
+	/* transition to long mode. along the way, we enable SSE. */
+
+	movl %cr4, %eax			/* enable PAE and SSE */
+	orl $(CR4_PAE | CR4_OSFXSR), %eax
+	movl %eax, %cr4
+	clts
+
+	movl $pml4, %eax		/* load page base */
+	movl %eax, %cr3
+
+	movl $X86_EFER_MSR, %ecx	/* enable long mode */
+	rdmsr
+	orl $X86_EFER_MSR_LME, %eax
+	wrmsr
+
+	movl %cr0, %eax			/* enable paging */
+	orl $CR0_PG, %eax
+	movl %eax, %cr0
+
+	/* jump into long mode, reload the segment registers (again). */
+
+	jmpl $X86_KERNEL_CS_64, $1f
+.code64
+1:	movl $X86_KERNEL_DS_64, %eax
+	movw %ax, %ds
+	movw %ax, %es
+	movw %ax, %ss
+	movw %ax, %fs
+	movw %ax, %gs
+
+	movl $X86_KERNEL_TSS, %eax
+	ltr %ax
+
+	cld
+	xorl %eax, %eax
+	movq $__bss_start, %rdi
+	movq $__bss_num_qwords, %rcx
+	rep stosq
+
+#ifdef CONFIG_X86_MULTIBOOT_INFO
+	/* If EBX isn't 0, then we have a valid multiboot info struct. */
+
+	testl %ebx, %ebx
+	jz 1f
+	movl %ebx, %esi
+
+	movl $__X86_MULTIBOOT_INFO_SIZEOF, %ecx
+	movq $x86_multiboot_info, %rdi
+	rep
+	movsb
+1:
+#endif /* CONFIG_X86_MULTIBOOT_INFO */
+
+	/*
+	 * set up SSE in case something uses the floating-point unit during
+	 * early initialization (either directly, or if GCC gets clever)
+	 */
+
+	xorl %edi, %edi
+	call x86_sse_init
+
+	/* don't replace CALL with JMP; honor the ABI stack alignment! */
+
+	call z_cstart
+
+stop:	jmp stop
+
+/*
+ * void x86_sse_init(struct k_thread *thread);
+ *
+ * Initialize floating-point state to something sane. If 'thread' is
+ * not NULL, then the resulting FP state is saved to thread->arch.sse.
+ */
+
+.global x86_sse_init
+x86_sse_init:
+	fninit
+	ldmxcsr mxcsr
+	testq %rdi, %rdi
+	jz 1f
+	fxsave _thread_offset_to_sse(%rdi)
+1:	retq
+
+mxcsr:	.long X86_MXCSR_SANE
+
+/*
+ * FIXME: The multiboot header is identical (for obvious reasons) to the
+ * version in ia32/crt0.S. They should be refactored into a common file.
+ */
+
+#ifdef CONFIG_X86_MULTIBOOT
+	.align 4
+	.long X86_MULTIBOOT_HEADER_MAGIC
+	.long X86_MULTIBOOT_HEADER_FLAGS
+	.long -(X86_MULTIBOOT_HEADER_MAGIC + X86_MULTIBOOT_HEADER_FLAGS)
+#ifdef CONFIG_X86_MULTIBOOT_FRAMEBUF
+	.fill 5,4,0				/* (unused exec layout) */
+	.long 0					/* linear graphics mode */
+	.long CONFIG_X86_MULTIBOOT_FRAMEBUF_X	/* width */
+	.long CONFIG_X86_MULTIBOOT_FRAMEBUF_Y	/* height */
+	.long 32				/* depth */
+#endif /* CONFIG_X86_MULTIBOOT_FRAMEBUF */
+#endif
+
+/*
+ * XXX: describe __swap, __resume, stacks
+ */
+
+.globl _k_neg_eagain	/* from errno.c: int _k_neg_eagain = -EAGAIN; */
+
+.globl __swap
+__swap:
```

Comment:
Don't we need to use the new switch() stuff in order to eventually implement SMP?

Reply:
Yep, in the scheduler's current form.  I am currently using `__swap()` because the 32-bit subarch does.  We can add the `z_arch_switch()` stuff later when we need SMP, preferably to both subarches at the same time, or we can rework the SMP scheduler to work with `__swap()`.

Reply:
> Maybe I'm missing some context, but why would we want to keep both of these?? As far as I'm concerned, once this is up and running with SMP and memory protection we can and absolutely should remove x86_64 arch

I agree, but see the war that nearly broke out in #17385 because I didn't use enough `xuk` and `x86_64` code in this patch: @andyross said he wasn't going to approve this until it basically had feature-parity with his `x86_64` implementation (including SMP).  Rather than wasting time explaining all the reasons why that was a bad idea, I decided to side-step the issue by repositioning this patch.  I see no reason why we need to eliminate `x86_64` any time soon.  (Of course, I will have no hand in maintaining it, either.)

Reply:
> We can add the z_arch_switch() stuff later when we need SMP, preferably to both subarches at the same time, or we can rework the SMP scheduler to work with __swap().

OK fine w/me. No objections to implementing SMP in a later PR, rather than this one. I would want to see it use switch, so we are using the same design for all SMP arches (Xtensa, very soon ARC).

No objections with leaving this current swap implementation here meanwhile.

> I agree, but see the war that nearly broke out in #17385 because I didn't use enough xuk and x86_64 code in this patch: @andyross said he wasn't going to approve this until it basically had feature-parity with his x86_64 implementation (including SMP). Rather than wasting time explaining all the reasons why that was a bad idea, I decided to side-step the issue by repositioning this patch. I see no reason why we need to eliminate x86_64 any time soon. (Of course, I will have no hand in maintaining it, either.)

We're gonna need to hash this stuff out, whether in github, a meeting, or whatever. Given that we're intending to have this arch target SMP and memory protection working for 2.1, assuming we succeed I would want to see the old x86_64 removed in 2.1.